subsys/prefetch/src/main/java/org/commonjava/indy/subsys/prefetch/PrefetchWorker.java

```diff
@@ -0,0 +1,145 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.commonjava.indy.subsys.prefetch;
+
+import org.commonjava.indy.model.core.RemoteRepository;
+import org.commonjava.maven.galley.TransferException;
+import org.commonjava.maven.galley.TransferManager;
+import org.commonjava.maven.galley.model.ConcreteResource;
+import org.commonjava.maven.galley.model.ListingResult;
+import org.commonjava.maven.galley.model.Transfer;
+import org.commonjava.maven.galley.util.PathUtils;
+import org.slf4j.Logger;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class PrefetchWorker
+        implements Runnable
+{
+    private TransferManager transfers;
+
+    private PrefetchManager prefetchManager;
+
+    private PrefetchFrontier frontier;
+
+    private Map<RemoteRepository, List<ConcreteResource>> resources;
+
+    private final Logger logger;
+
+    private static final String LISTING_HTML_FILE = "index.html";
+
+    public PrefetchWorker( final TransferManager transfers, final PrefetchFrontier frontier,
+                           Map<RemoteRepository, List<ConcreteResource>> resources,
+                           final PrefetchManager prefetchManager, final Logger logger )
+    {
+        this.transfers = transfers;
+        this.frontier = frontier;
+        this.resources = resources;
+        this.prefetchManager = prefetchManager;
+        this.logger = logger;
+    }
+
+    @Override
+    public void run()
+    {
+        if ( resources == null || resources.isEmpty() )
+        {
+            logger.info( "No resources for downloading" );
+            return;
+        }
+
+        logger.info( "Start downloading: {}", resources );
+        final AtomicBoolean scheduled = new AtomicBoolean( false );
+        for ( Map.Entry<RemoteRepository, List<ConcreteResource>> entry : resources.entrySet() )
+        {
+            final RemoteRepository repo = entry.getKey();
+            final List<ConcreteResource> res = entry.getValue();
+            res.forEach( r -> {
+                try
+                {
+                    final String path = r.getPath();
+                    if ( path == null || path.equals( "" ) || path.endsWith( "/" ) || path.endsWith(
+                            LISTING_HTML_FILE ) )
+                    {
+                        List<String> dirPaths = buildPaths( r );
+                        logger.info( "{} is folder, will use it to schedule new Resources {}", r, dirPaths );
+                        frontier.scheduleRepo( repo, dirPaths );
+                        scheduled.set( true );
+                    }
+                    else
+                    {
+                        logger.info( "{} is file", r );
+                        final Transfer tr = transfers.retrieve( r );
+                        if ( !exists( tr ) )
```

Comment:
Do we need to wait for the download to complete before doing this? I think retrieve just returns when the download starts...

Reply:
Not sure if the retrieve is working to just return immediately(I remember that there is a download job in that method). But sure, we do not need care if it exists or not in the prefetch. Will remove it.